### ISSUE TITLE
ALOps App Publish to onPrem with Eset Security fails

### ISSUE THREAD

**Issue by JohnnyUndercover** [HodorNV/ALOps#235]
We are using an Build Agent at our customer to publish apps with the command "ALOps App Publish". On the Build Agent there is an Antivirus application running [(Eset Security)](https://www.eset.com/)
Until today everything worked pretty well 👍 But the antivirus "Eset Security" got an update and we received an error.
**First Error**
```
.9734973Z ##[section]Starting: Publish PRI-BC-VisualSignalModul
2020-10-26T13:43:29.9837477Z ==============================================================================
2020-10-26T13:43:29.9837808Z Task         : ALOps App Publish
2020-10-26T13:43:29.9838084Z Description  : Publish AL Extension to Business Central
2020-10-26T13:43:29.9838313Z Version      : 1.435.1864
2020-10-26T13:43:29.9838493Z Author       : Hodor
2020-10-26T13:43:29.9838740Z Help         : Publish Business Central extension to service tier.
2020-10-26T13:43:29.9839046Z ==============================================================================
2020-10-26T13:43:31.2426014Z *** Validate configuration
2020-10-26T13:43:31.4217944Z ##[error]In Zeile:1 Zeichen:1
+ ([rUntiMe.intERoPsERViCES.MArShal]::([ruNTimE.iNtEROpSeRVIces.mARSHaL ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Das Skript enthält schädliche Daten und wurde von Ihrer Antivirensoftware blockiert.
2020-10-26T13:43:31.4454429Z ##[section]Finishing: Publish PRI-BC-VisualSignalModul
```

I found a solutin for this in another Issue #108 
After setting the variable `alops_disable_amsi: true` i got a success message in devops, but the app was not updated in the database.

We removed the  `alops_disable_amsi: true` variable again and also got the error from above again. (With a new release)
Then we disabled the AMSI feature at the antivirus software and got tried again with a new release. But this time we got the follwing error:

**Second Error**
```
2020-10-26T12:50:27.6459981Z ##[section]Starting: Publish PRI-BC-VisualSignalModul
2020-10-26T12:50:27.6561670Z ==============================================================================
2020-10-26T12:50:27.6561978Z Task         : ALOps App Publish
2020-10-26T12:50:27.6562217Z Description  : Publish AL Extension to Business Central
2020-10-26T12:50:27.6562440Z Version      : 1.435.1864
2020-10-26T12:50:27.6562668Z Author       : Hodor
2020-10-26T12:50:27.6562919Z Help         : Publish Business Central extension to service tier.
2020-10-26T12:50:27.6563226Z ==============================================================================
2020-10-26T12:50:31.1028070Z *** Validate configuration
2020-10-26T12:50:31.4477267Z ##[error]Das Argument "0" mit dem Wert  "2971776705384" für "GetStartComSlot" kann nicht in den Typ "System.Type" konvertiert werden: "Der Wert "2971776705384" vom Typ "System.IntPtr" kann nicht in den Typ "System.Type" konvertiert werden."
2020-10-26T12:50:31.4809866Z ##[section]Finishing: Publish PRI-BC-VisualSignalModul
```

**Our Yaml**
```
steps:
- task: Hodor.hodor-alops.alopspublishtask.ALOpsAppPublish@1
  displayName: 'Publish PRI-BC-VisualSignalModul'
  inputs:
    nav_serverinstance: 'BC_TEST'
    nav_artifact_app_filter: '*Visual Signal Module*.app'
```

Any suggestions how to solve this?

**Comment by waldo1001** [comment]
You need to disable virus scan, indeed - either with the "alops_disable_amsi" setting, or simply removing the anti-virus.

Can you please give the log of the pipeline where you disabled the virus in the pipeline?
Did you try alops.externaldeployer?

**Comment by JohnnyUndercover** [comment]
The log when "alops_disable_amsi" is set to true:
```
2020-10-26T15:38:40.4370544Z ##[section]Starting: Publish PRI-BC-VisualSignalModul
2020-10-26T15:38:40.4475623Z ==============================================================================
2020-10-26T15:38:40.4475921Z Task         : ALOps App Publish
2020-10-26T15:38:40.4476160Z Description  : Publish AL Extension to Business Central
2020-10-26T15:38:40.4476414Z Version      : 1.435.1864
2020-10-26T15:38:40.4476590Z Author       : Hodor
2020-10-26T15:38:40.4476836Z Help         : Publish Business Central extension to service tier.
2020-10-26T15:38:40.4477168Z ==============================================================================
2020-10-26T15:38:41.2671346Z *** Configure AMSI for session: disable (in memory)
2020-10-26T15:38:41.2894912Z 0
2020-10-26T15:38:41.8503082Z *** Validate configuration
2020-10-26T15:38:42.1082401Z *** For documentation, please visit   : https://www.alops.be/documentation
2020-10-26T15:38:42.1086300Z 
2020-10-26T15:38:42.2573548Z ##[section]Finishing: Publish PRI-BC-VisualSignalModul
```

`Did you try alops.externaldeployer?`
No, what is it and how can try it?

**Comment by waldo1001** [comment]
That's an odd output :(.  Can you share screenshots or settings?

About the external deployer, you can find more on my blog: https://www.waldo.be/2020/06/15/deploying-from-devops-the-right-way-enabling-external-deployment-in-onprem-business-central-environments/

**Comment by JohnnyUndercover** [comment]
sure:
![image](https://user-images.githubusercontent.com/30835996/97271277-a7582280-1830-11eb-80cb-8f2f6d85810d.png)
![image](https://user-images.githubusercontent.com/30835996/97271294-ac1cd680-1830-11eb-93c3-6f67ec320ee2.png)
![image](https://user-images.githubusercontent.com/30835996/97271315-b212b780-1830-11eb-8fb7-5c30d3f0ab18.png)

I also looked into the event log at the build agent and cleand the _work directory and stop/started the build agent service. But it changed nothing.

I recognized that there is still somthing blocked by the anti virus
![image](https://user-images.githubusercontent.com/30835996/97272847-c3f55a00-1832-11eb-943e-c830f6480dc7.png)

I'm pretty sure this is the problem.

For the external deployer the api needs to be accessible from outside right?

**Comment by JohnnyUndercover** [comment]
Solved the Problem with adding an exclusion for this specific event (hash) to the antivirus.
![image](https://user-images.githubusercontent.com/30835996/97280187-39b1f380-183c-11eb-81f1-7646e4226242.png)

Maybe this will help someone else ;)

**Comment by waldo1001** [comment]
Thanks!

About the external deployer: indeed, you would be able to push to publish to ANY NST from a central build agent (you don't need a DevOps agent on each NST).  Still, it's going to execute powershell in the background (that's what included in alops.externaldeployer), so I'm not sure the virus scanner will be a problem there.. .

but yeah, if a program doesn't allow you to execute powershell, there is not much we can do about that, other than set the ansi-parameter, but the rest is up to that program .. .

thanks for the solution
